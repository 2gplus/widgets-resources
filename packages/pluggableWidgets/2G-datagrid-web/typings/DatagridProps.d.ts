/**
 * This file was generated from Datagrid.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { ComponentType, CSSProperties, ReactNode } from "react";
import { ActionValue, DynamicValue, EditableValue, ListValue, ListActionValue, ListAttributeValue, ListExpressionValue, ListWidgetValue, WebIcon } from "mendix";
import { Big } from "big.js";

export type SortingTypeEnum = "local" | "remote";

export type ShowContentAsEnum = "attribute" | "dynamicText" | "customContent";

export type HidableEnum = "yes" | "hidden" | "no";

export type WidthEnum = "autoFill" | "autoFit" | "manual";

export type AlignmentEnum = "left" | "center" | "right";

export interface ColumnsType {
    showContentAs: ShowContentAsEnum;
    attribute?: ListAttributeValue<string | Big | boolean | Date>;
    content?: ListWidgetValue;
    dynamicText?: ListExpressionValue<string>;
    header?: DynamicValue<string>;
    sortProperty: string;
    filter?: ReactNode;
    sortable: boolean;
    resizable: boolean;
    draggable: boolean;
    hidable: HidableEnum;
    width: WidthEnum;
    size: number;
    alignment: AlignmentEnum;
    columnClass?: ListExpressionValue<string>;
}

export type PaginationEnum = "buttons" | "virtualScrolling";

export type PagingPositionEnum = "bottom" | "top" | "hide";

export type ShowEmptyPlaceholderEnum = "none" | "custom";

export type DefaultTriggerEnum = "singleClick" | "doubleClick";

export type SelectionModeEnum = "single" | "multi";

export type RenderModeEnum = "link" | "button";

export type ButtonStyleEnum = "default" | "inverse" | "primary" | "info" | "success" | "warning" | "danger";

export interface ButtonsType {
    caption: string;
    action?: ListActionValue;
    icon?: DynamicValue<WebIcon>;
    renderMode: RenderModeEnum;
    buttonStyle: ButtonStyleEnum;
}

export interface FilterListType {
    filter: ListAttributeValue<string | Big | boolean | Date>;
}

export interface ColumnsPreviewType {
    showContentAs: ShowContentAsEnum;
    attribute: string;
    content: { widgetCount: number; renderer: ComponentType<{ caption?: string }> };
    dynamicText: string;
    header: string;
    sortProperty: string;
    filter: { widgetCount: number; renderer: ComponentType<{ caption?: string }> };
    sortable: boolean;
    resizable: boolean;
    draggable: boolean;
    hidable: HidableEnum;
    width: WidthEnum;
    size: number | null;
    alignment: AlignmentEnum;
    columnClass: string;
}

export interface ButtonsPreviewType {
    caption: string;
    action: {} | null;
    icon: { type: "glyph"; iconClass: string; } | { type: "image"; imageUrl: string; } | null;
    renderMode: RenderModeEnum;
    buttonStyle: ButtonStyleEnum;
}

export interface FilterListPreviewType {
    filter: string;
}

export interface DatagridContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex?: number;
    advanced: boolean;
    datasource: ListValue;
    sortingType: SortingTypeEnum;
    sortAttribute?: EditableValue<string>;
    sortAscending?: EditableValue<boolean>;
    onSortChangedAction?: ActionValue;
    columns: ColumnsType[];
    columnsFilterable: boolean;
    pageSize: number;
    pagination: PaginationEnum;
    pagingPosition: PagingPositionEnum;
    showEmptyPlaceholder: ShowEmptyPlaceholderEnum;
    emptyPlaceholder?: ReactNode;
    rowClass?: ListExpressionValue<string>;
    defaultTrigger: DefaultTriggerEnum;
    onTrigger?: ListActionValue;
    selectionMode: SelectionModeEnum;
    tableLabel?: DynamicValue<string>;
    buttons: ButtonsType[];
    columnsSortable: boolean;
    columnsResizable: boolean;
    columnsDraggable: boolean;
    columnsHidable: boolean;
    configurationAttribute?: EditableValue<string>;
    onConfigurationChange?: ActionValue;
    showHeaderFilters: boolean;
    filterList: FilterListType[];
    filtersPlaceholder?: ReactNode;
    filterSectionTitle?: DynamicValue<string>;
}

export interface DatagridPreviewProps {
    class: string;
    style: string;
    advanced: boolean;
    datasource: {} | { type: string } | null;
    sortingType: SortingTypeEnum;
    sortAttribute: string;
    sortAscending: string;
    onSortChangedAction: {} | null;
    columns: ColumnsPreviewType[];
    columnsFilterable: boolean;
    pageSize: number | null;
    pagination: PaginationEnum;
    pagingPosition: PagingPositionEnum;
    showEmptyPlaceholder: ShowEmptyPlaceholderEnum;
    emptyPlaceholder: { widgetCount: number; renderer: ComponentType<{ caption?: string }> };
    rowClass: string;
    defaultTrigger: DefaultTriggerEnum;
    onTrigger: {} | null;
    selectionMode: SelectionModeEnum;
    tableLabel: string;
    buttons: ButtonsPreviewType[];
    columnsSortable: boolean;
    columnsResizable: boolean;
    columnsDraggable: boolean;
    columnsHidable: boolean;
    configurationAttribute: string;
    onConfigurationChange: {} | null;
    showHeaderFilters: boolean;
    filterList: FilterListPreviewType[];
    filtersPlaceholder: { widgetCount: number; renderer: ComponentType<{ caption?: string }> };
    filterSectionTitle: string;
}
