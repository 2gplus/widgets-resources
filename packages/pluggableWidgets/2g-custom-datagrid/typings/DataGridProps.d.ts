/**
 * This file was generated from Datagrid.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { ComponentType, CSSProperties, ReactNode } from "react";
import {
    ActionValue,
    DynamicValue,
    EditableValue,
    ListValue,
    ListActionValue,
    ListAttributeValue,
    ListExpressionValue,
    ListWidgetValue
} from "mendix";
import { Big } from "big.js";

export type ShowContentAsEnum = "attribute" | "dynamicText" | "customContent";

export type HidableEnum = "yes" | "hidden" | "no";

export type WidthEnum = "autoFill" | "autoFit" | "manual";

export type AlignmentEnum = "left" | "center" | "right";

export interface ColumnsType {
    showContentAs: ShowContentAsEnum;
    attribute?: ListAttributeValue<string | Big | boolean | Date>;
    content?: ListWidgetValue;
    dynamicText?: ListExpressionValue<string>;
    header?: DynamicValue<string>;
    sortable: boolean;
    resizable: boolean;
    draggable: boolean;
    hidable: HidableEnum;
    width: WidthEnum;
    size: number;
    alignment: AlignmentEnum;
    columnClass?: ListExpressionValue<string>;
}

export type PaginationEnum = "buttons" | "virtualScrolling";

export type PagingPositionEnum = "bottom" | "top";

export type ShowEmptyPlaceholderEnum = "none" | "custom";

export type SelectionModeEnum = "single" | "multi";

export type DefaultTriggerEnum = "singleClick" | "doubleClick";

export interface ColumnsPreviewType {
    showContentAs: ShowContentAsEnum;
    attribute: string;
    content: { widgetCount: number; renderer: ComponentType<{ caption?: string }> };
    dynamicText: string;
    header: string;
    sortable: boolean;
    resizable: boolean;
    draggable: boolean;
    hidable: HidableEnum;
    width: WidthEnum;
    size: number | null;
    alignment: AlignmentEnum;
    columnClass: string;
}

export interface DatagridContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex?: number;
    advanced: boolean;
    datasource: ListValue;
    columns: ColumnsType[];
    pageSize: number;
    pagination: PaginationEnum;
    pagingPosition: PagingPositionEnum;
    showEmptyPlaceholder: ShowEmptyPlaceholderEnum;
    emptyPlaceholder?: ReactNode;
    rowClass?: ListExpressionValue<string>;
    selectionMode: SelectionModeEnum;
    defaultTrigger: DefaultTriggerEnum;
    onTrigger?: ListActionValue;
    columnsSortable: boolean;
    columnsResizable: boolean;
    columnsDraggable: boolean;
    columnsHidable: boolean;
    configurationAttribute?: EditableValue<string>;
    onConfigurationChange?: ActionValue;
}

export interface DatagridPreviewProps {
    class: string;
    style: string;
    advanced: boolean;
    datasource: {} | { type: string } | null;
    columns: ColumnsPreviewType[];
    pageSize: number | null;
    pagination: PaginationEnum;
    pagingPosition: PagingPositionEnum;
    showEmptyPlaceholder: ShowEmptyPlaceholderEnum;
    emptyPlaceholder: { widgetCount: number; renderer: ComponentType<{ caption?: string }> };
    rowClass: string;
    selectionMode: SelectionModeEnum;
    defaultTrigger: DefaultTriggerEnum;
    onTrigger: {} | null;
    columnsSortable: boolean;
    columnsResizable: boolean;
    columnsDraggable: boolean;
    columnsHidable: boolean;
    configurationAttribute: string;
    onConfigurationChange: {} | null;
}
