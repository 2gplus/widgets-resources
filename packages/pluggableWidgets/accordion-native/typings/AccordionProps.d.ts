/**
 * This file was generated from Accordion.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { ComponentType, ReactNode } from "react";
import { DynamicValue, NativeIcon } from "mendix";

export type CollapsibleEnum = "collapsibleNo" | "collapsibleYesExpanded" | "collapsibleYesCollapsed";

export interface AccordionProps<Style> {
    name: string;
    style: Style[];
    headerContent?: ReactNode;
    content: ReactNode;
    showHeader: boolean;
    collapsible: CollapsibleEnum;
    iconCollapsed?: DynamicValue<NativeIcon>;
    iconExpanded?: DynamicValue<NativeIcon>;
}

export interface AccordionPreviewProps {
    class: string;
    style: string;
    headerContent: { widgetCount: number; renderer: ComponentType<{caption?: string}> };
    content: { widgetCount: number; renderer: ComponentType<{caption?: string}> };
    showHeader: boolean;
    collapsible: CollapsibleEnum;
    iconCollapsed: { type: "glyph"; iconClass: string; } | { type: "image"; imageUrl: string; } | null;
    iconExpanded: { type: "glyph"; iconClass: string; } | { type: "image"; imageUrl: string; } | null;
}
