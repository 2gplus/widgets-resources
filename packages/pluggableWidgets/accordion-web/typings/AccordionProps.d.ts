/**
 * This file was generated from Accordion.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { ComponentType, CSSProperties, ReactNode } from "react";
import { DynamicValue } from "mendix";

export type HeaderRenderModeEnum = "text" | "custom";

export interface GroupsType {
    headerRenderMode: HeaderRenderModeEnum;
    headerText: DynamicValue<string>;
    headerContent?: ReactNode;
}

export interface GroupsPreviewType {
    headerRenderMode: HeaderRenderModeEnum;
    headerText: string;
    headerContent: { widgetCount: number; renderer: ComponentType<{caption?: string}> };
}

export interface AccordionContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex?: number;
    groups: GroupsType[];
}

export interface AccordionPreviewProps {
    class: string;
    style: string;
    groups: GroupsPreviewType[];
}
