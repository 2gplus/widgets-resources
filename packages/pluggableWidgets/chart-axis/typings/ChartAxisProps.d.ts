/**
 * This file was generated from ChartAxis.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
export interface ChartAxisProps<Style> {
    name: string;
    style: Style[];
}

export interface ChartAxisPreviewProps {
    class: string;
    style: string;
}
