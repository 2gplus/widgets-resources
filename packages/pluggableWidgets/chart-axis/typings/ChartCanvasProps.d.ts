/**
 * This file was generated from ChartCanvas.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { ComponentType, ReactNode } from "react";

export interface ChartCanvasProps<Style> {
    name: string;
    style: Style[];
    content?: ReactNode;
}

export interface ChartCanvasPreviewProps {
    class: string;
    style: string;
    content: { widgetCount: number; renderer: ComponentType };
}
