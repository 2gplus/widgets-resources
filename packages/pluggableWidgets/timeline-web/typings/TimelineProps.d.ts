/**
 * This file was generated from Timeline.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { ComponentType, CSSProperties } from "react";
import {
    DynamicValue,
    ListValue,
    ListActionValue,
    ListAttributeValue,
    ListExpressionValue,
    ListWidgetValue,
    WebIcon
} from "mendix";

export type RenderModeEnum = "basic" | "custom";

export type GroupByKeyEnum = "day" | "month" | "year";

export type GroupByDayOptionsEnum = "dayName" | "dayMonth" | "fullDate";

export type GroupByMonthOptionsEnum = "month" | "monthYear";

export interface TimelineContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex: number;
    renderMode: RenderModeEnum;
    data: ListValue;
    eventTime: ListAttributeValue<Date>;
    title?: ListAttributeValue<string>;
    description?: ListAttributeValue<string>;
    time?: ListExpressionValue<string>;
    icon?: DynamicValue<WebIcon>;
    groupByKey: GroupByKeyEnum;
    showGroupHeader: boolean;
    groupByDayOptions: GroupByDayOptionsEnum;
    groupByMonthOptions: GroupByMonthOptionsEnum;
    customIcon?: ListWidgetValue;
    customGroupHeader?: ListWidgetValue;
    customTitle?: ListWidgetValue;
    customEventDateTime?: ListWidgetValue;
    customDescription?: ListWidgetValue;
    onClick?: ListActionValue;
}

export interface TimelinePreviewProps {
    class: string;
    style: string;
    renderMode: RenderModeEnum;
    data: {} | null;
    eventTime: string;
    title: string;
    description: string;
    time: string;
    icon: { type: "glyph"; iconClass: string } | { type: "image"; imageUrl: string } | null;
    groupByKey: GroupByKeyEnum;
    showGroupHeader: boolean;
    groupByDayOptions: GroupByDayOptionsEnum;
    groupByMonthOptions: GroupByMonthOptionsEnum;
    customIcon: { widgetCount: number; renderer: ComponentType };
    customGroupHeader: { widgetCount: number; renderer: ComponentType };
    customTitle: { widgetCount: number; renderer: ComponentType };
    customEventDateTime: { widgetCount: number; renderer: ComponentType };
    customDescription: { widgetCount: number; renderer: ComponentType };
    onClick: {} | null;
}
