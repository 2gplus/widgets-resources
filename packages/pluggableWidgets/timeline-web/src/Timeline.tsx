import { createElement, ReactElement, ReactNode, useMemo } from "react";
import { TimelineContainerProps } from "../typings/TimelineProps";
import "./ui/Timeline.scss";
import { ActionValue, WebIcon } from "mendix";
import TimelineComponent, { getGroupHeaderByType } from "./components/TimelineComponent";

export interface BasicItemType {
    icon?: WebIcon;
    title?: string;
    eventDateTime?: string;
    description?: string;
    action?: ActionValue;
}

export interface CustomItemType {
    icon?: ReactNode;
    title?: ReactNode;
    eventDateTime?: ReactNode;
    description?: ReactNode;
    action?: ActionValue;
    groupHeader?: ReactNode;
}

export type ItemType = BasicItemType | CustomItemType;

export default function Timeline(props: TimelineContainerProps): ReactElement {
    const groupedEvents = useMemo((): Map<string, ItemType[]> => {
        const eventsMap = new Map<string, ItemType[]>();

        props.data.items?.forEach(item => {
            let constructedItem: ItemType;
            const eventTime = props.eventTime?.(item);
            const date = eventTime?.value;
            let groupKey;

            if (props.renderMode === "basic") {
                const headerOption =
                    props.groupByKey === "day"
                        ? props.groupByDayOptions
                        : props.groupByKey === "month"
                        ? props.groupByMonthOptions
                        : "year";

                groupKey = getGroupHeaderByType(eventTime?.formatter, headerOption, date);
                constructedItem = {
                    icon: props.icon?.value,
                    title: props.title?.(item)?.displayValue,
                    eventDateTime: props.time?.(item)?.value,
                    description: props.description?.(item)?.displayValue,
                    action: props.onClick?.(item)
                };
            } else {
                groupKey = getGroupHeaderByType(eventTime?.formatter, props.groupByKey, date);
                constructedItem = {
                    icon: props.customIcon?.(item),
                    groupHeader: props.customGroupHeader?.(item),
                    title: props.customTitle?.(item),
                    eventDateTime: props.customEventDateTime?.(item),
                    description: props.customDescription?.(item),
                    action: props.onClick?.(item)
                };
            }

            const currentDates = eventsMap.get(groupKey);
            if (currentDates) {
                currentDates.push(constructedItem);
            } else {
                eventsMap.set(groupKey, [constructedItem]);
            }
        });

        return eventsMap;
    }, [props.data]);

    return (
        <TimelineComponent
            name={props.name}
            data={groupedEvents}
            showGroupHeader={props.showGroupHeader}
            renderMode={props.renderMode}
            eventOrder={props.eventOrder}
        />
    );
}
