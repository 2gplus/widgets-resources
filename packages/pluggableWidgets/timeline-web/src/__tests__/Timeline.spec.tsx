import { shallow } from "enzyme";
import { createElement } from "react";
import { TimelineContainerProps } from "../../typings/TimelineProps";
import { dynamicValue, EditableValueBuilder, ListValueBuilder } from "@widgets-resources/piw-utils";
import Timeline from "../Timeline";

jest.mock("mendix/components/web/Icon", () => jest.requireActual("./__mocks__/mendix/components/web/Icon"));

describe("Timeline", () => {
    const listValueBuilder = ListValueBuilder();
    const defaultProps: TimelineContainerProps = {
        name: "timeline",
        class: "test",
        tabIndex: 0,
        renderMode: "basic",
        groupByKey: "day",
        data: listValueBuilder.simple(),
        eventTime: () => new EditableValueBuilder<Date>().withValue(new Date(1453, 4, 29)).build(),
        showGroupHeader: true,
        title: () => new EditableValueBuilder<string>().withValue("title").build(),
        description: () => new EditableValueBuilder<string>().withValue("description").build(),
        time: () => dynamicValue<string>("time"),
        customIcon: () => <img src={"test"} />,
        customGroupHeader: () => <p>Custom Divider</p>,
        customTitle: () => <p>Custom Title</p>,
        customEventDateTime: () => <p>Custom Event Date Time</p>,
        customDescription: () => <p>Custom Description</p>,
        groupByMonthOptions: "month",
        groupByDayOptions: "dayName",
        eventOrder: "descending"
    };
    it("renders correctly with basic props", () => {
        const component = shallow(<Timeline {...defaultProps} />);
        expect(component).toMatchSnapshot();
    });
});
