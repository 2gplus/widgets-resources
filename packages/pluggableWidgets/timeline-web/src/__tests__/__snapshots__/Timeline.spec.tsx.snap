// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline renders correctly with basic props 1`] = `
<TimelineComponent
  data={
    Map {
      "Formatted Sun May 29 1453 00:00:00 GMT+0019 (Central European Summer Time)" => Array [
        Object {
          "action": undefined,
          "description": "Formatted description",
          "eventDateTime": "time",
          "icon": undefined,
          "title": "Formatted title",
        },
        Object {
          "action": undefined,
          "description": "Formatted description",
          "eventDateTime": "time",
          "icon": undefined,
          "title": "Formatted title",
        },
      ],
    }
  }
  name="timeline"
  renderMode="basic"
  showGroupHeader={true}
/>
`;
